api/api-pipeline-config-v7/src/main/java/com/thoughtworks/go/apiv7/admin/pipelineconfig/PipelineConfigControllerV7.java

```diff
@@ -0,0 +1,242 @@
+
+    @Autowired
+    public PipelineConfigControllerV7(PipelineConfigService pipelineConfigService, ApiAuthenticationHelper apiAuthenticationHelper, EntityHashingService entityHashingService, PasswordDeserializer passwordDeserializer, GoConfigService goConfigService, GoConfigPluginService goConfigPluginService) {
+        super(ApiVersion.v7);
+        this.pipelineConfigService = pipelineConfigService;
+        this.apiAuthenticationHelper = apiAuthenticationHelper;
+        this.entityHashingService = entityHashingService;
+        this.passwordDeserializer = passwordDeserializer;
+        this.goConfigService = goConfigService;
+        this.goConfigPluginService = goConfigPluginService;
+    }
+
+    @Override
+    public String etagFor(PipelineConfig pipelineConfig) {
+        return entityHashingService.md5ForEntity(pipelineConfig);
+    }
+
+    @Override
+    public PipelineConfig doFetchEntityFromConfig(String name) {
+        return pipelineConfigService.getPipelineConfig(name);
+    }
+
+    @Override
+    public PipelineConfig buildEntityFromRequestBody(Request req) {
+        ConfigHelperOptions options = new ConfigHelperOptions(goConfigService.getCurrentConfig(), passwordDeserializer);
+        JsonReader jsonReader = GsonTransformer.getInstance().jsonReaderFrom(req.body());
+        if ("PUT".equalsIgnoreCase(req.requestMethod())) {
+            return PipelineConfigRepresenter.fromJSON(jsonReader, options);
+        }
+        return PipelineConfigRepresenter.fromJSON(jsonReader.readJsonObject("pipeline"), options);
+    }
+
+    @Override
+    public Consumer<OutputWriter> jsonWriter(PipelineConfig pipelineConfig) {
+        return writer -> PipelineConfigRepresenter.toJSON(writer, pipelineConfig);
+    }
+
+    @Override
+    public String controllerBasePath() {
+        return Routes.PipelineConfig.BASE;
+    }
+
+    @Override
+    public void setupRoutes() {
+        path(controllerPath(), () -> {
+            before("", mimeType, this::setContentType);
+            before("/*", mimeType, this::setContentType);
+            before("", this::verifyContentType);
+            before("/*", this::verifyContentType);
+            before("", mimeType, apiAuthenticationHelper::checkPipelineCreationAuthorizationAnd403);
+            before(Routes.PipelineConfig.NAME, mimeType, apiAuthenticationHelper::checkPipelineGroupAdminUserAnd403);
+            before(Routes.PipelineConfig.EXPORT, mimeType, apiAuthenticationHelper::checkPipelineGroupAdminUserAnd403);
+
+            post("", mimeType, this::create);
+
+            get(Routes.PipelineConfig.NAME, mimeType, this::show);
+            put(Routes.PipelineConfig.NAME, mimeType, this::update);
+            delete(Routes.PipelineConfig.NAME, mimeType, this::destroy);
+
+            get(Routes.PipelineConfig.EXPORT, mimeType, this::export);
+
+            exception(RecordNotFoundException.class, this::notFound);
+        });
+    }
+
+    public String export(Request req, Response res) {
```

Comment:
This method/endpoint is the "real" change here to `v7`; the rest is basically copy-pasted from `v6`, which accounts for a lot of lines in this PR.